tests/nn/decoding/expected_risk_minimization_test.py

```diff
@@ -0,0 +1,103 @@
+        scores = [score for state in states for score in state.score]
+        return SimpleDecoderState(batch_indices, action_histories, scores)
+
+
+class SimpleDecoderStep(DecoderStep[SimpleDecoderState]):
+    @overrides
+    def take_step(self,
+                  state: SimpleDecoderState,
+                  max_actions: int = None,
+                  allowed_actions: List[Set] = None) -> List[SimpleDecoderState]:
+        if allowed_actions is None:
+            # For each element in the group, the allowed actions are adding 1 or 2 to the last
+            # element.
+            allowed_actions = [{1, 2} for _ in state.batch_indices]
+        indexed_next_states: Dict[int, List[SimpleDecoderState]] = defaultdict(list)
+        for batch_index, action_history, score, actions in zip(state.batch_indices,
+                                                               state.action_history,
+                                                               state.score,
+                                                               allowed_actions):
+
+            for action in actions:
+                next_item = action_history[-1] + action
+                new_history = action_history + [next_item]
+                # For every action taken, we reduce the score by 1.
+                new_score = score - 1
+                new_state = SimpleDecoderState([batch_index],
+                                               [new_history],
+                                               [new_score])
+                indexed_next_states[batch_index].append(new_state)
+        next_states: List[SimpleDecoderState] = []
+        for batch_next_states in indexed_next_states.values():
+            if max_actions is not None:
+                batch_next_states = batch_next_states[:max_actions]
+            next_states.extend(batch_next_states)
+        return next_states
+
+
+class TestExpectedRiskMinimization(AllenNlpTestCase):
+    def setUp(self):
+        super(TestExpectedRiskMinimization, self).setUp()
+        self.initial_state = SimpleDecoderState([0], [[0]], [Variable(torch.Tensor([0.0]))])
+        self.decoder_step = SimpleDecoderStep()
+        # Cost is the number of odd elements in the action history.
+        self.supervision = lambda state: Variable(torch.Tensor([sum([x%2 != 0 for x in
+                                                                     state.action_history[0]])]))
+        # High beam size ensures exhaustive search.
+        self.trainer = ExpectedRiskMinimization(beam_size=100,
+                                                normalize_by_length=False,
+                                                max_decoding_steps=10)
+
+    def test_get_finished_states(self):
+        finished_states = self.trainer._get_finished_states(self.initial_state, self.decoder_step)
+        state_info = []
+        for state in finished_states:
+            state_info.append((state.action_history[0], int(state.score[0].data)))
```

Comment:
`state_info = [(state.action_history[0], int(state.score[0].data)) for state in finished_states]`, instead of your for loop.